### ISSUE TITLE
Language server issues in outline pane

### ISSUE THREAD

**Issue by eafpres** [spyder-ide/spyder#14384]
## Problem Description

Several "symptoms" involving the outline pane; I'm unsure how best to document, so looking for direction:
1) The outline often collapses. e.g. I expand all, then when working in the code it collapses.
2) The server busy spinner (next to the dropdown in the pane) randomly starts spinning. Sometimes this is accompanied by the editor and console hanging

### What steps reproduce the problem?

1. Open a project
2. Work in some code

### What is the expected output? What do you see instead?

Outline retains state (what is collapsed or not remains collapsed or not)
UI does not hang with spinning server

### Paste Traceback/Error Below (if applicable)

```python-traceback

<many more of these messages>
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...
WARNING:spyder.plugins.completion.languageserver.client:The send queue is full! Retrying...

<a few of these messages>
QProcess: Destroyed while process ("c:\\users\\bbate\\envs\\keras-gpu-6\\scripts\\pythonw.exe") is still running.

```

## Versions

```
Python 3.7.4 64-bit  
Qt 5.12.9  
PyQt5 5.12.3  
Windows 10 
```

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  3.0.4 (OK)
cloudpickle >=0.5.0             :  1.3.0 (OK)
diff_match_patch >=20181111     :  20181111 (OK)
intervaltree >=3.0.2            :  3.1.0 (OK)
IPython >=4.0                   :  7.19.0 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
nbconvert >=4.0                 :  5.6.1 (OK)
numpydoc >=0.6.0                :  0.9.2 (OK)
paramiko >=2.4.0                :  2.7.2 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.7.3 (OK)
pygments >=2.0                  :  2.5.2 (OK)
pylint >=1.0                    :  1.9.4 (OK)
pyls >=0.36.1;<1.0.0            :  0.36.1 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.1.1             :  0.2.1 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  0.7.3 (OK)
qtconsole >=4.7.7               :  4.7.7 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
setuptools >=39.0.0             :  41.0.0 (OK)
sphinx >=0.6.6                  :  1.8.5 (OK)
spyder_kernels >=1.10.0;<1.11.0 :  1.10.0 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog >=0.10.3               :  0.10.3 (OK)
zmq >=17                        :  19.0.2 (OK)

# Optional:
cython >=0.21                   :  0.29.14 (OK)
matplotlib >=2.0.0              :  3.3.2 (OK)
numpy >=1.7                     :  1.19.3 (OK)
pandas >=1.1.1                  :  1.1.4 (OK)
scipy >=0.17.0                  :  1.4.1 (OK)
sympy >=0.7.3                   :  None (NOK)

```

**Comment by ccordoba12** [comment]
Hey @eafpres, thanks for reporting. You said:

> The outline often collapses. e.g. I expand all, then when working in the code it collapses.

Could you post a simple, reproducible example of that? Otherwise is hard for us to know how to tackle this problem.

> The server busy spinner (next to the dropdown in the pane) randomly starts spinning. Sometimes this is accompanied by the editor and console hanging

I'll try to come up with something for our next version (4.2.1), to be released next week. Otherwise, I'll take a look at it for sure for 4.2.2.

**Comment by eafpres** [comment]
Minimum repro code:

```
# -*- coding: utf-8 -*-
"""
Created on Fri Dec 11 12:18:17 2020

@author: bbate
"""
# %% block 1
#
def foo(x):
    result = x**2
    return result
#
# %% block 2
#
def bar(y):
    result = y / 2
    return result
#
# %% block 3
#
def fubar(z, x, y):
    result = z + x * 10 - y
    return result
#
# %% main code
#
x = 10
y = 20
z = 30
print(fubar(z, foo(x), bar(y)))
```

Attached video shows the outline collapse on editing the code
1) Expand the outline
2) put cursor on line 29 (assignment to z)
3) hit return

outline collapses

No errors in the internal console for this issue
The other issue (hanging, internal server errors) I have not found a simple repro yet.

[IMG_1735.zip](https://github.com/spyder-ide/spyder/files/5681235/IMG_1735.zip)

**Comment by ccordoba12** [comment]
Hey @eafpres, about the problems you reported:

> The outline often collapses. e.g. I expand all, then when working in the code it collapses.

I can't reproduce this error in our latest `4.x` branch. So please wait until you can install Spyder 4.2.2 and let us know if the error is still there.

> The server busy spinner (next to the dropdown in the pane) randomly starts spinning. Sometimes this is accompanied by the editor and console hanging

I just merged PR #13864, which should help with that. Again, please test it once 4.2.2 is out.

**Comment by eafpres** [comment]
> please test it once 4.2.2 is out.

Okay, will do. Thank you.

**Comment by CAM-Gerlach** [comment]
@eafpres 4.2.2 is out, FYI

**Comment by eafpres** [comment]
> eafpres 4.2.2 is out, FYI

As far as I can tell, the collapsing issue is solved (actually at least from 4.2.1).
The spinning issue is there, but I need to attempt some kind of reproducible example in 4.2.2
If anything new / else appears, I'll open a new issue.
More to come, and as always, thanks.

**Comment by eafpres** [comment]
So when I updated to 4.2.2 a code I'm working on (3200++ lines in the editor)  started crashing constantly. This was not happening in 4.2.1. I found that if I commented out a bunch of code analysis errors like so:
```
# pylint: disable=invalid-name, too-many-lines, line-too-long, wrong-import-position, wrong-import-order, ungrouped-imports, no-name-in-module, no-member, unidiomatic-typecheck, import-outside-toplevel, protected-access, self-assigning-variable, too-many-branches

```
The problem in 4.2.2 goes away.  

It does "appear" the spinning thing in the (long) outline is "better" in 4.2.2 than 4.2.1.  However, I think there are issue when there are a lot of outline elements or the code is large or something along those lines.

Right now I can't use 4.2.2 due to another issue I submitted-error dialog popping up constantly in the editor. In 4.2.1, the outline gets confused often as to exactly where in the code things are, so I have to not use it when the spinning thing is there, or it will go to some place other than where I want to go.  I'm hoping that's more reliable in 4.2.2, and look forward to testing that more.

**Comment by CAM-Gerlach** [comment]
Sorry you're running into these issues! Strange; disabling Pylint errors _shouldn't_ have any effect, since Spyder doesn't use Pylint for real-time linting, only on-demand. The fact that you're seeing one, unless there's something else going on here, could possibly be indicative of Python Language Server using the Pylint backend for linting even when its never used, since pylint is really slow compared to pyflakes/pycodestyle (which Spyder uses), slowing things down considerably on longer/more complex files.

Could you describe how you updated Spyder from 4.2.1 to 4.2.2? With Anaconda (default channel or conda forge?), with the installer, or another method? Based on what you've told us, it seems likely the change was caused by an upgrade to the PyLS or PyLS-Spyder version, as opposed to Spyder itself—if you're able, can you downgrade to 4.2.1 without modifying any other dependencies (or the minimum possible) and check if the problem is still present? Finally, is this problem the same as you've reported on #14835 , and if not, does it occur to the same copy of Spyder (i.e. so your troubleshooting output there would be valid here)? Thanks!

**Comment by eafpres** [comment]
To be clear, I don't really know what is going on, but I observed this change.  The popup error thing goes away with exactly the same code files (same project) just exiting Spyder, running pip install spyder==4.2.1 in Powershell, and restarting Spyder.  

So I think I have done what you asked--only changed the Spyder version and the error went away.  The issue I raised as not being above to completely verify this issue in 4.2.2 is indeed reported in 14835; I wanted that separate as I think it is (not sure).

I don't know what is happening when the outline panel is spinning, other than it does not know about code changes and may go to the wrong line in the code if I click on the outline.

I will attempt to repro this behavior regarding commenting out pylint errors when I have a bit of time.  I realize my code has a lot of moving pieces and this could be subtle.

**Comment by ccordoba12** [comment]
> So when I updated to 4.2.2 a code I'm working on (3200++ lines in the editor) started crashing constantly

What do you mean exactly by this? What are the crashes you're observing?

**Comment by eafpres** [comment]
> > So when I updated to 4.2.2 a code I'm working on (3200++ lines in the editor) started crashing constantly
> 
> What do you mean exactly by this? What are the crashes you're observing?

The entire application suddenly terminates.  I can use the recovery file depending on the circumstances. Because of the way it crashes, I have no internal console etc. I can try to formulate a repro, when I get more time.

**Comment by CAM-Gerlach** [comment]
@eafpres Okay, I assume you don't get any relevant output in the Powershell console window you ran it from? Can you try running in debug mode, with `spyder --debug-info verbose`?

**Comment by eafpres** [comment]
> I assume you don't get any relevant output in the Powershell console window you ran it from? Can you try running in debug mode, with `spyder --debug-info verbose`?

Correct. No Powershell output. I’ll get back to repro this once I’m past a Wednesday project milestone.

**Comment by eafpres** [comment]
Today I tested with my large (ish) code (3000+ lines) in 4.2.1.
Changing the pylint comment seems to have no effect; my observation may have been coincidental. I also tested re-running the code analysis without the items commented out, and this had no effect on performance. 
Running in debug mode produced no messages when the outline panel was spinning (or any messages)
I confirm that if I make changes, the spinner starts
It "seems" (hard to validate) that if the change is flagged as an error, the spinning lasts longer
During spinning, if the # of code lines was changed, jumping to a location using the outline goes to the wrong place; i.e. the number of lines or however the outline syncs to the code becomes un-matched
